osquery/tables/events/darwin/openbsm_events.cpp

```diff
@@ -156,10 +197,109 @@ Status OpenBSMExecVESubscriber::Callback(
     }
   }
   r["uptime"] = INTEGER(tables::getUptime());
+
+  try {
+    r["parent"] = INTEGER(ppid_map.at(pid));
+  } catch (std::out_of_range) {
+    /* Mapping doesn't exist so no fork was captured. Not fatal. Ignoring. */
+  }
+
   add(r);
   return Status(0, "OK");
 }
 
+Status OpenBSMProcEventsSubscriber::handleFork(
+    const OpenBSMEventContextRef& ec) {
+  uint32_t ppid = 0; // Parent PID
+  uint32_t cpid = 0; // Child PID
+
+  for (const auto& tok : ec->tokens) {
+    switch (tok.id) {
+      /* The parent is the process issuing the syscall. Its PID is given on the
+       * subject token.
+       */
+      case AUT_SUBJECT32:
+        ppid = tok.tt.subj32.pid;
+        break;
+      case AUT_SUBJECT64:
+        ppid = tok.tt.subj64.pid;
+        break;
+      case AUT_SUBJECT32_EX:
+        ppid = tok.tt.subj32_ex.pid;
+        break;
+      case AUT_SUBJECT64_EX:
+        ppid = tok.tt.subj64_ex.pid;
+        break;
+
+      /* Child PID is given as the first argument here. This could also be
+       * extract from the return token as the child PID should be the return
+       * value of the fork syscall given to the parent.
+       */
+      case AUT_ARG32:
+        if (tok.tt.arg32.no == 0) {
+          cpid = tok.tt.arg32.val;
+        }
+        break;
+      case AUT_ARG64:
+        if (tok.tt.arg64.no == 0) {
+          cpid = tok.tt.arg64.val;
+        }
+        break;
+
+      /* Check whether fork succeeded. Upon failure stop processing as no child
+       * is created.
+       */
+      case AUT_RETURN32:
+        if (tok.tt.ret32.status != 0) {
+          return Status(0);
+        }
+        break;
+      case AUT_RETURN64:
+        if (tok.tt.ret64.err != 0) {
+          return Status(0);
+        }
+    }
+  }
+
+  /* If we succeeded to to capture both parent and child PIDs add them to the
+   * mapping. This information should always be available, if not the event is
+   * malformed.
+   */
+  if (ppid != 0 && cpid != 0) {
+    ppid_map[cpid] = ppid;
```

Comment:
If we had a process that was starting thousands of child processes, this could get large. Do you think it's worth implementing a maximum size for this map? If it is, the size of it should be configured by a flag.

Reply:
See above.